app/WebParsing/ReqParser.hs

```diff
@@ -135,14 +135,47 @@ singleParser = do
     num <- Parsec.count 3 Parsec.digit
     -- TODO: Make the last two letters more restricted.
     sess <- Parsec.count 2 Parsec.alphaNum
-    return $ J (code ++ num ++ sess)
+    return $ J (code ++ num ++ sess) ""
+
+-- | Parser for single courses or "atomic" Reqs represented by a J.
+justParser :: Parser Req
+justParser = do
+    code <- Parsec.count 3 Parsec.letter
+    num <- Parsec.count 3 Parsec.digit
+    sess <- Parsec.count 2 Parsec.alphaNum
+    Parsec.spaces
+    lpar <- Parsec.option "" $ Parsec.string "("
+    tmp <- if lpar == "" then return (Right "") else markInfoParser
+    case tmp of
+        Left mark -> return $ GRADE mark $ J (code ++ num ++ sess) ""
+        Right info -> return $ J (code ++ num ++ sess) info
+    where
+    markInfoParser = do
+        grade <- Parsec.try (percentParser <|> letterParser <|> infoParser)
+        _ <- Parsec.string ")"
+        return grade
+            where
+            percentParser = do
```

Comment:
These are now duplicated from above; so you should make the top-level helper functions.